### PR TITLE
Adds find and update routes to the template

### PR DESCRIPTION
**Pull request by abbasmousavi**
Adds find and update routes to the template

### CODE REVIEW

File: Sources/App/Controllers/TodoController.swift [hunk 1]
```diff
@@ -20,4 +20,23 @@ final class TodoController {
...
+    }
+
+    /// Finds a parameterized `Todo`.
+    func find(_ req: Request) throws -> Future<Todo> {
+
+        do {
+            let todo = try req.parameter(Todo.self)
+            return todo
+        } catch {
+            throw Abort(.notFound)
```
Comment (tanner0101) [inline Sources/App/Controllers/TodoController.swift]: vapor should be serving a 404 for this automatically

File: Sources/App/Controllers/TodoController.swift [hunk 2]
```diff
@@ -20,4 +20,23 @@ final class TodoController {
             return todo.delete(on: req)
         }.transform(to: .ok)
     }
+
+    /// Updates a decoded `Todo` to the database.
+    func update(_ req: Request) throws -> Future<Todo> {
+
```
Comment (tanner0101) [inline Sources/App/Controllers/TodoController.swift]: you should use the `router.put(Todo.self, ...)` syntax here :)

File: Sources/App/Controllers/TodoController.swift (hunk 1 above)
Comment (abbasmousavi) [reply Sources/App/Controllers/TodoController.swift]: Currently it returns a 500.
the error is `[ ERROR ] FluentError.modelNotFound: No model with ID 7 was found (Model.swift:265)`
 I have removed do-catch.

File: Sources/App/Controllers/TodoController.swift (hunk 2 above)
Comment (abbasmousavi) [reply Sources/App/Controllers/TodoController.swift]: I don't get the point here.

**Comment by abbasmousavi** [comment]
@tanner0101 

What is your opinion about adding a generic controller for simple CRUD paths to api template like this

```
import Vapor
import Foundation
import FluentSQLite

/// Controlers basic CRUD operations on `Todo`s.
final class Controller <T:SQLiteModel & Parameter & Content> {
    
    
    /// Returns a list of all `Todo`s.
    func index(_ req: Request) throws -> (Future<[T]>) {
        return T.query(on: req).all()
    }

    /// Saves a decoded `Todo` to the database.
    func create(_ req: Request) throws -> Future<T> {
        return try req.content.decode(T.self).flatMap(to: T.self) { liveStream in
            return liveStream.save(on: req)
        }
    }

    /// Deletes a parameterized `Todo`.
    func delete(_ req: Request) throws -> Future<HTTPStatus> {
        return try (req.parameter(T.self) as! EventLoopFuture<T>).flatMap(to: Void.self) { liveStream in
            return liveStream.delete(on: req)
        }.transform(to: .ok)
    }

    func update(_ req: Request) throws -> Future<T> {

        return try req.content.decode(T.self).flatMap(to: T.self) { liveStream in
            return liveStream.update(on: req)
        }
    }

    func item(_ req: Request) throws -> Future<T> {

        do {
            let post = try req.parameter(T.self) as! EventLoopFuture<T>
            return post
        } catch {
            throw Abort(.notFound)
        }
    }
```

**Comment by abbasmousavi** [comment]
@tanner0101

What is your opinion about adding a generic controller for simple CRUD paths to api template like this
```
import Vapor
import Foundation
import FluentSQLite

/// Controlers basic CRUD operations on `Todo`s.
protocol RESTController: :RouteCollection {
    associatedtype T: PostgreSQLModel & Parameter & Content
}

extension RESTController {

   func boot(router: Router) throws {
        router.get(use: index)
        router.get(T.parameter, use: find)
        router.post(use: create)
        router.put(T.parameter, use: update)
        router.delete(T.parameter, use: delete)
    
    }
    
    /// Returns a list of all `Todo`s.
    func index(_ req: Request) throws -> (Future<[T]>) {
        return T.query(on: req).all()
    }

    /// Saves a decoded `Todo` to the database.
    func create(_ req: Request) throws -> Future<T> {
        return try req.content.decode(T.self).flatMap(to: T.self) { model in
            return model.save(on: req)
        }
    }

    /// Deletes a parameterized `Todo`.
    func delete(_ req: Request) throws -> Future<HTTPStatus> {
        return try (req.parameter(T.self) as! EventLoopFuture<T>).flatMap(to: Void.self) { model in
            return model.delete(on: req)
        }.transform(to: .ok)
    }

    func update(_ req: Request) throws -> Future<T> {

        return try req.content.decode(T.self).flatMap(to: T.self) { model  in
            return model.update(on: req)
        }
    }

    func find(_ req: Request) throws -> Future<T> {

            return try req.parameter(T.self) as! EventLoopFuture<T>
    }
```

then for every model just 

```
final class TodoController: RESTController {
    
    typealias T = Todo
}
```
and the router code will be 

```
try router.grouped("todos").register(collection: TodoController())
```

This will simplify making simple REST backends for everyone.

**Comment by 0xTim** [comment]
Tanner did have a RESTControlller protocol or implementation planned, similar to how Vapor 2 did it. Not sure what the plan is for that?

As for the update method, it should really be a `PUT` method to `/todo/ID` to update a single ID. For REST a `PUT` method to the resource (i.e. just `/todo`) should update the whole collection

**Comment by abbasmousavi** [comment]
@0xTim, I agree, a RESTController protocol is a good idea. I have edited the proposed code to be a protocol.

As for the update method, Is it enough to change
`router.put("todos", use: todoController.update)`
to
`router.put("todos", Todo.parameter, use: todoController.update)` ?

**Comment by pedantix** [comment]
I would suggest considering two types in the rest controller

DTO/Model


But that is just me

**Comment by 0xTim** [comment]
Closing as this repo has been superseded by vapor/template - feel free to reopen on that repository if this PR still makes sense. Thanks for contributing!

**Pull request closed without merging**